### ISSUE TITLE
Blender / Luxcore not responsive when rendering in larger resolutions.

### ISSUE THREAD

**Issue by theogott** [LuxCoreRender/BlendLuxCore#521]
This is a simple issue. If I would talk as a programmer I would say "There are some DoEvents() missing".
The point is taht if you render low-resoulution below 4000x4000 Pixel, this is NO ISSUE.
Luxcore reacts, you can change the window position and size, and even press escape to stop rendering.
But if you like to render with higher resolution, lets say 8000x8000 or higher then Luxcore/Blender does not respond to any user action for MINUTES.
SO maybe you can check where the "DoEvents() are missing?
I have a 24 Core CPU with 2x GTX 1080 that render together on 128 GB RAM.
SO i guess its not the hardware.

**Comment by Dade916** [comment]
You have enough CPU ram but are you sure to have enough GPU ram ? The number of GPUs doesn't matter because each GPU has its own private local copy of the data.

Are you using no-tile rendering ? For that kind of ultra-high resolution you should enable tile rendering.

**Comment by theogott** [comment]
It does not matter what kind of scene i am rendering. Even just taking a Plane and the default cube at 8000x8000 you should see the effect. Also it does not happen with for example Cycles or radeon render. Its a resolution specific problem. And i wonder if you tell me that you do not have it. I have 2 computers, and both have that effect: Higher resolution - worse responsiveness. 
Just try to render something at 8000x8000 it should happen anyway. No matter what scene.
Did you render something (whatever)  at high resolution yourself and the problem does not appear?

**Comment by Dade916** [comment]
Are you using no-tile rendering ? For that kind of ultra-high resolution you should enable tile rendering.

**Comment by theogott** [comment]
Yes i do that partly. But then there is no caustics. So can not use it anywhere. 
Also even the tiled rendering - while it is more stable (can run 2 at the same time at low prio)
can not render everything. Besides that the tiled rendering is more responsive - yes - but its not as responsive as it should. 
It still can take 30 Seconds before anything moves.

**Comment by Dade916** [comment]
PhotonGI caustic cache will allow you to render caustics even with tile rendering.

Is Blender-only to not be responsive or all the OS ?

**Comment by theogott** [comment]
Ok, good to know. This is just Blender. Other Programs around are very responsive. Also it only happens with Luxrender, all other engines like Evee, Cycles, or Radeon Render are responsive. And because it only happens and increases with hi-resolutions, i think that this can not be much. Maybe there is some sort of DoEvent() between then drawing of the frame?
I wonder if you do not have that effect at all?

**Comment by Theverat** [comment]
I can reproduce the problem with 8000x8000 render resolution.

**Comment by Theverat** [comment]
I've done some print debugging and this is what I get with an 8000x8000 resolution:
```
update_stats() took 0.000 s
beginning engine.framebuffer.draw()
     engine.begin_result() took 0.151 s
     _convert_combined() took 1.545 s
     engine.end_result() took 21.710 s
engine.framebuffer.draw() took 23.407 s

update_stats() took 0.000 s
beginning engine.framebuffer.draw()
     engine.begin_result() took 0.147 s
     _convert_combined() took 1.500 s
     engine.end_result() took 21.676 s
engine.framebuffer.draw() took 23.323 s
```
Most of the time is spent in `engine.end_result()`, which is a Blender API method: https://docs.blender.org/api/current/bpy.types.RenderEngine.html?highlight=end_result#bpy.types.RenderEngine.end_result
I'll have to check if we are using this method wrong somehow, if not, it might be a performance problem in Blender.

**Comment by juangea** [comment]
As a workaround maybe a kind of “headless” node could be implemented? 

Like disabling frame buffer output until the engine ends the render or something like that

**Comment by theogott** [comment]
Also, I have tested other Renderengines like Radeon Pro Render and its absolutely responsive at any time.
Therefore i doubt its an error in Blender.
I expect that there are some kind of "DoEvents()" that must be called between screen refreshes to react on Blender-Inputs.
This looks like it it done only between complete Screen-Refreshes currently.
Which is going to be a rising problem with resolution. 
If this will be called all ... 250 Scan-Lines or like that or with a timer it could make things as responsive as are all other rendering engines. I would recommend to check that program-part that will receive messages from Blender-Input (for example somebody presses ESCAPE Key). This part must be called more often,. maybe with a timer.

**Comment by juangea** [comment]
Have you tried to increase the time between refreshes of the frame buffer?

**Comment by Theverat** [comment]
> I expect that there are some kind of "DoEvents()" that must be called between screen refreshes to react on Blender-Inputs.

The problem is that the `end_result()` method is one contiguous block of execution time and can't be broken up from outside.
Anyway, let me first check if we are calling `end_result()` in some suboptimal way.

**Comment by theogott** [comment]
I think you are on the way. 
If needed, just update not the full resolution but blocks like others.
This can start from Resolutions above 2K for example that you update in 2K Blocks.
Just an idea, may be you have a much better.
Anyway responsiveness is important for the user experience.

**Comment by Theverat** [comment]
> I tested your Cycles Scene it doies not hang here in any way.

Are you sure? This is a recording of how it looks on my end: https://youtu.be/YYL3iCxCs2w

**Comment by theogott** [comment]
I see. Maybe because you used CPU Rendering, using the GPU it hangs a bit at the start and then while rendering it is fine. 
I could not take that video, have to restart my computer first. 
But with 2xGTX 1080, Cycles while rendering your testing scene is responsive here.

Here is a video that shows the other issue first (it just happened while i prepared ...).
With the testing Scene from above.
https://youtu.be/EOHMrUuaXUs

**Comment by Theverat** [comment]
You are rendering with many small tiles, not one single 8000x8000 tile.

**Comment by theogott** [comment]
You are right because Cycles renders faster in tiles.
Anyway even if i render with Tiles Luxrender is not responsive as it could.
If we talk about tiles. try Radeon Pro Render. They have all CPU-Threads and the CPU always render in ONE TILE.
I think that has a lot advantages about the Caching Memory and you have never the problem that the CPU-Tiles are not ready and block the GPU from Rendering. Also about responsiveness. RadeonPro is the best Candidate. Its always perfect responsive wether with Tiles or not.
